### PR TITLE
feat: Add unfold command

### PR DESCRIPTION
**Pull request by hudclark**
# Description

> [!NOTE]
> This PR description originally used examples where the `generator` closure returned a list. It has since been updated to use records instead.

The `unfold` command allows users to dynamically generate streams of data. The stream is generated by repeatedly invoking a `generator` closure. The `generator` closure accepts a single argument and returns a record containing two optional keys: 'out' and 'next'. Each invocation, the 'out' value, if present, is added to the stream. If a 'next' key is present, it is used as the next argument to the closure, otherwise generation stops.

The name "unfold" is borrowed from other functional-programming languages. Whereas `fold` (or `reduce`) takes a stream of values and outputs a single value, `unfold` takes a single value and outputs a stream of values.

### Examples

A common example of using `unfold` is to generate a fibbonacci sequence. See [here](https://github.com/rust-itertools/itertools/blob/6ffdac103cf72dfd3b62a4de6dc25440b942e473/src/sources.rs#L65)  for an example of this in rust's `itertools`.

```nushell
> unfold [0, 1] {|fib| {out: $fib.0, next: [$fib.1, ($fib.0 + $fib.1)]} } | first 10
───┬────
 0 │  0
 1 │  1
 2 │  1
 3 │  2
 4 │  3
 5 │  5
 6 │  8
 7 │ 13
 8 │ 21
 9 │ 34
───┴────
```

This command is particularly useful when consuming paginated APIs, like Github's. Previously, nushell users might use a loop and buffer responses into a list, before returning all responses at once. However, this behavior is not desirable if the result result is very large. Using `unfold` avoids buffering and allows subsequent pipeline stages to use the data concurrently, as it's being fetched.

#### Before
```nushell
mut pages = []
for page in 1.. {
  let resp = http get (
    {
      scheme: https,
      host: "api.github.com",
      path: "/repos/nushell/nushell/issues",
      params: {
	page: $page,
	per_page: $PAGE_SIZE
      }
    } | url join)

  $pages = ($pages | append $resp)

  if ($resp | length) < $PAGE_SIZE {
    break
  }
}
$pages
```

#### After
```nu
unfold 1 {|page|
  let resp = http get (
    {
      scheme: https,
      host: "api.github.com",
      path: "/repos/nushell/nushell/issues",
      params: {
	page: $page,
	per_page: $PAGE_SIZE
      }
    } | url join)

  if ($resp | length) < $PAGE_SIZE {
    {out: $resp}
  } else {
    {out: $resp, next: ($page + 1)}
  }
}
```

# User-Facing Changes

- An `unfold` generator is added to the default context.

# Tests + Formatting

# After Submitting

Given the complexity of the `generator` closure's return value, it would be good to document the semantics of `unfold` and provide some in-depth examples showcasing what it can accomplish.

### CODE REVIEW

**Review by bobhy — COMMENTED** [review]
Potential perf issue having to clone all generated return data?

File: crates/nu-command/src/generators/unfold.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+                // no data -> output nothing and stop.
+                Ok(PipelineData::Empty) => (None, None),
+
+                Ok(PipelineData::Value(Value::List { vals, .. }, ..)) => match vals.len() {
+                    // [] -> output nothing and stop.
+                    0 => (None, None),
+                    // [a] -> output `a` and stop.
+                    1 => (Some(vals[0].clone()), None),
+                    // [a b] -> output `a` and continue with `b`.
+                    2 => (Some(vals[0].clone()), Some(vals[1].clone())),
```
Comment (bobhy) [inline crates/nu-command/src/generators/unfold.rs:149]: Potential perf issue having to clone all generated data?

**Comment by bobhy** [comment]
Generally seems a concise and pretty elegant command for Nu to have.
But I don't think you had to go to this effort to get paginated results streamed from a service.  It seems that emitting a result from a loop body implicitly makes them into a stream.  Here's an updated github example demonstrating the effect:

```
> cat test.nu

def get_pages [num_pages: int = 5, page_size = 3 ] {
    $env.my_github_fg_pat = "Bearer github_pat_..."

    1..$num_pages | each {|page| 

        let resp = http get (
            {
                scheme: https,
                host: "api.github.com",
                path: "/repos/nushell/nushell/issues",
                params: {
                    page: $page,
                    per_page: $page_size
                }
            } | url join) --headers [Authorization $env.my_github_fg_pat]

        let trunc_resp = ($resp | select number title user.login state )

        print -e $"emitting page ($page)"
    
        $trunc_resp
    }
}
```
Runniing it:
```
> source test.nu
> get_pages | describe
emitting page 1
emitting page 2
emitting page 3
emitting page 4
emitting page 5
list<table<number: int, title: string, user_login: string, state: string>> (stream)
# seems to be streaming results

> get_pages | each { |r| print -e $"       processing ($r.number)"} 
emitting page 1
       processing [10492, 10489, 10488]
emitting page 2
       processing [10486, 10484, 10483]
emitting page 3
       processing [10480, 10479, 10478]
emitting page 4
       processing [10476, 10468, 10465]
emitting page 5
       processing [10464, 10459, 10454]
╭────────────╮
│ empty list │
╰────────────╯
# seems to show the loop body getting each page of results without waiting for them to be batched up

```

**Comment by hudclark** [comment]
> Generally seems a concise and pretty elegant command for Nu to have. But I don't think you had to go to this effort to get paginated results streamed from a service. It seems that emitting a result from a loop body implicitly makes them into a stream. Here's an updated github example demonstrating the effect:

Appreciate the feedback @bobhy!

I agree that `unfold` isn't a requirement for an API where the page ordering known ahead of time. However, I don't think we can accomplish the same for cursor-based APIs ([example](https://slack.engineering/evolving-api-pagination-at-slack/)), which I tend to interact with most. I don't have a publicly accessible API handy, but the code would look like this:
```nushell
# On the first invocation, the cursor is null
unfold null {|page|
  let resp = http get (
    {
      scheme: https,
      host: "<example>",
      path: "/path",
      params: {
	cursor: $cursor,
	per_page: $PAGE_SIZE
      }
    } | url join)

  # Each call, the response contains the next cursor to use or null to stop
  if resp.cursor? == null {
    [$resp]
  } else {
    [$resp, $resp.cursor]
  }
}
```

Trying to accomplish the above using a loop or `each`  (and keeping streaming semantics), users will run into: `capture of mutable variable`.

**Comment by bobhy** [comment]
Ah, I get it!  You need a mutable variable that you can set at the bottom
of the closure that retains its value at the top of the next invocation of
the closure.
You can't use a capture because Nu complains about capture of a mutable
value, and you can't use an environment variable because it gets reset to
initial value at the top of the loop.

At least, I ran into both of those issues in my sample.

So maybe this kind of enhancement is the only alternative if we can't relax
either of the above behaviors.

**Comment by hudclark** [comment]
> Ah, I get it! You need a mutable variable that you can set at the bottom of the closure that retains its value at the top of the next invocation of the closure. You can't use a capture because Nu complains about capture of a mutable value, and you can't use an environment variable because it gets reset to initial value at the top of the loop. At least, I ran into both of those issues in my sample. So maybe this kind of enhancement is the only alternative if we can't relax either of the above behaviors.

That's exactly it! And while it's very useful in this specific example, it has a wide variety of use-cases. It's a common building-block in the FP libraries/languages I've come across, but it's not always named "unfold". 

In Kotlin, it's called [`generateSequence`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.sequences/generate-sequence.html). In Reactive Extensions, it's called [`genereate`](https://rxjs.dev/api/index/function/generate). In languages/libraries that have a strong FP background, I usually see it called `unfold`.

Given "`unfold`" would be new to many nushell users, a name like `generate` or similar may be more approachable and easier to discover when searching for the command.

**Comment by hudclark** [comment]
> however, i just wanted to note that the output of this is not the same as the output of the `mut` script you'd have to use right now 😮
> * the former, i.e. with `mut` returns a table with 1005 rows
> * the latter, i.e. with `unfold` returns a list of tables, 10 have 100 rows and the last one has 5 of them, giving a total of 1005 rows
> 
> Note adding a little `flatten` after `unfold` solves this difference it appears
> 
> was that on purpose? 😋

Good call out! You're right that the two scripts are slightly different. The closure in `unfold` can output exactly zero or one value each invocation. Since we're fetching a page of results each invocation (and our page size is greater than one), the result is a stream of pages. Like you suggested, `flatten` can be used to get a stream of individual items instead.

> also @hudclark, i'm still not sure i understand why this command is called `unfold` 😕

@amtoine  In most libraries/languages I'm familiar with, this function is named `unfold`. The name comes from it being effectively the inverse of the `fold` function. I'll link a few examples below:
- [Scala](https://www.scala-lang.org/api/2.13.6/scala/collection/Iterator$.html#unfold[A,S](init:S)(f:S=%3EOption[(A,S)]):Iterator[A])
- [Rambda.js](https://ramdajs.com/docs/#unfold)
- [Haskell](https://hoogle.haskell.org/?hoogle=unfoldr)
- [Rust](https://docs.rs/unfold/latest/unfold/)

A number of blogs and articles can be found for searching "unfold functional programming" or similar. One such example: https://medium.com/earnest-engineering/unfolding-paginated-apis-195f5608bcc9

That being said, as I mentioned earlier, some languages/libraries use a name like `generate` instead, although this seems less common.

File: crates/nu-command/src/generators/unfold.rs (hunk 1 above)
Comment (hudclark) [reply crates/nu-command/src/generators/unfold.rs]: Nice callout. I pushed a fix that avoids the copying :)

**Comment by amtoine** [comment]
@hudclark 
thanks a ton for the explaination, i see now :pray: 

i'm fine with `unfold` or `generate` then :yum:

File: crates/nu-command/tests/commands/unfold.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    let actual = nu!("unfold 0 {|x| [a b c]}");
+    assert!(actual.err.contains("Invalid block return"));
+}
+
+#[test]
+fn unfold_allows_nothing_state() {
+    let actual = nu!(pipeline(
+        r#"
+        unfold 0 {|x|
+          if $x == $nothing {
```
Comment (amtoine) [inline crates/nu-command/tests/commands/unfold.rs:35]: because of https://github.com/nushell/nushell/pull/10478, could you replace all `$nothing`s with `null`? :yum:

**Review by amtoine — CHANGES_REQUESTED** [review]
i like the look and feel of the GitHub paging, but i'm in love with this example
```nushell
unfold [0, 1] {|fib| [$fib.0, [$fib.1, ($fib.0 + $fib.1)]] } | first 20
```
:heart: 

i just have a remark about the tests but otherwise this looks good and fun to me :ok_hand:

File: crates/nu-command/tests/commands/unfold.rs (hunk 2 above)
Comment (hudclark) [reply crates/nu-command/tests/commands/unfold.rs]: Fixed

Comment (amtoine) [reply crates/nu-command/tests/commands/unfold.rs]: :pray:

File: crates/nu-command/src/generators/unfold.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    Category, Example, IntoInterruptiblePipelineData, IntoPipelineData, PipelineData, ShellError,
+    Signature, Span, Spanned, SyntaxShape, Type, Value,
+};
+
+#[derive(Clone)]
+pub struct Unfold;
+
+impl Command for Unfold {
+    fn name(&self) -> &str {
+        "unfold"
```
Comment (jakeswenson) [inline crates/nu-command/src/generators/unfold.rs:16]: would it make sense to have this under the `seq` command, as `seq unfold`? 

https://www.nushell.sh/commands/docs/seq.html#frontmatter-title-for-generators

File: crates/nu-command/tests/commands/unfold.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
+use nu_test_support::{nu, pipeline};
+
+#[test]
+fn unfold_single_element_break() {
+    let actual = nu!("unfold 1 {|x| if $x == 3 { [$x] } else { [$x, ($x + 1)] }} | to nuon");
```
Comment (jakeswenson) [inline crates/nu-command/tests/commands/unfold.rs:5]: Personally, I'm not in love with the syntax to using lists (as tuples) 

Not sure what I would like better though, records might be close:

`{ result: $x, state: ($x + 1) }`

**Review by jakeswenson — COMMENTED** [review]
A few thoughts

File: crates/nu-command/src/generators/unfold.rs (hunk 3 above)
Comment (hudclark) [reply crates/nu-command/src/generators/unfold.rs:16]: While I think `unfold` definitely belongs in the `generators` family, I don't think it belongs under `seq`. Seq's description is to "Output sequences of numbers.". `unfold` is a more general purpose tool. While many examples have focused around sequences of numbers, real-world use will likely be see a lot of paginating APIs and interfacing with external systems.

File: crates/nu-command/tests/commands/unfold.rs (hunk 4 above)
Comment (hudclark) [reply crates/nu-command/tests/commands/unfold.rs:5]: > Personally, I'm not in love with the syntax to using lists (as tuples)

@jakeswenson, I went back and forth here as well. To compare, here's a couple examples exploring how using records would look:
```nushell
unfold [0, 1] {|fib| {output: $fib.0, next: [$fib.1, ($fib.0 + $fib.1)]} } | first 10
```

```nushell
unfold 1 {|page|
  let resp = http get (
    {
      scheme: https,
      host: "api.github.com",
      path: "/repos/nushell/nushell/issues",
      params: {
	page: $page,
	per_page: $PAGE_SIZE
      }
    } | url join)

  if ($resp | length) < $PAGE_SIZE {
    {output: $resp}
  } else {
    {output: $resp, next: ($page + 1)}
  }
}
```

One reason I originally used a list are some corner cases around "invalid" returns:

If the generator returns a value that either isn't a record, or is a record that without an `output` key a runtime error would likely have to occur. 

```nushell
> unfold 1 {|x| if $x == 3 { $x } else { {output: $x, next: ($x + 1)} }}
<error stating invalid return. $x does not contain an `output` key>
```

If a `next` key existed, but no `output` key existed, I'm not sure whether that would indicate to stop, to output nothing and keep going, or to output `null` and keep going. Likewise, if any keys other than `output` or `next` are present in the return value, a runtime error would make sense.

--

All that being said, I agree that using lists in place of tuples isn't great. I'm open to changing to using records. It may be an improvement to the user experience and is also likely easier to document.

Comment (hudclark) [reply crates/nu-command/tests/commands/unfold.rs]: After thinking it over, I've gone forward with using records for the return type, rather than lists.

I think the command is easier to understand after the change.

**Comment by hudclark** [comment]
> [!IMPORTANT]
> The `unfold` command has been updated to return records rather than lists. The PR description has been updated to reflect the change.
> Use now looks like: `unfold [0, 1] {|fib| {out: $fib.0, next: [$fib.1, ($fib.0 + $fib.1)]} } | first 10`

File: crates/nu-command/src/generators/unfold.rs (hunk 3 above)
Comment (jakeswenson) [reply crates/nu-command/src/generators/unfold.rs:16]: That sounds like a documentation issue with `seq` being "for numbers" only - sequences are a general purpose concept. In my view, for nushell, unfold will always produce a sequence/list/table of zero or more items/rows.

So, since I consider sequences generalized it makes sense to group sequence generators under `seq`. Just my two cents. would love to hear others thoughts on this though.

**Review by amtoine — APPROVED** [review]
TL;DR: my review is that i like this new command :wink:

File: crates/nu-command/src/generators/unfold.rs (hunk 3 above)
Comment (hudclark) [reply crates/nu-command/src/generators/unfold.rs:16]: I'm open to hearing others input as well.

IMHO, I think of `seq` as a [math sequence](https://en.wikipedia.org/wiki/Sequence). Given unfold is a more general stream source/builder, I feel that it best belongs in the `generators` family of commands.

**Comment by fdncred** [comment]
Nice work! Let's run with this. If we have to change it, we can always do that too.

**Comment by NotTheDr01ds** [comment]
Beautiful!  Note that I predicted something like this [18 months ago](https://stackoverflow.com/a/72063355/11810933):

> Side-note: Personal opinion -- I envision that Nushell will eventually add a `yield` keyword to allow generator expressions. This will simply the above example further by allowing it inside a `reduce` that can accumulate the results.

But I like the streaming aspect of `unfold` even better.
